### PR TITLE
Add metadata about the sticker pack that a sticker image belongs to.

### PR DESCRIPTION
**Pull request by rxl881**
We need to add information about the sticker pack that sticker images belong so that users can discover integrations (and add / use them themselves).

### CODE REVIEW

File: event-schemas/examples/m.sticker [hunk 1]
```diff
@@ -15,7 +15,12 @@
       "w": 140,
       "size": 73602
     },
-    "url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP"
+    "url": "mxc://matrix.org/sHhqkFCvSkFwtmvtETOtKnLP",
+    "stickerpack": {
+      "name": "Rabbits",
+      "description": "Cute rabbit stickers",
+      "stickerpack_url": "https://scalar.vector.im/integration/stickerpicker/rabbits123"
```
Comment (turt2live) [inline event-schemas/examples/m.sticker]: What's the format of this URL? I've been working quite a lot in this area, so a formal spec of this 'sharing' would be appreciated, preferably with an opportunity to provide feedback.

**Comment by ara4n** [comment]
@rxl881 plz can haz an explanation of how the stickerpack_url is meant to work?

**Comment by turt2live** [comment]
Thanks for clarifying what the URL is supposed to be. 

Having the URL be a link to the integration manager introduces a compatibility problem between managers. For instance, if someone were to have a 3rd party sticker picker then there's a good chance that Scalar (for example) wouldn't be able to add the sticker pack. Likewise, a 3rd party integration manager won't know how to add a Scalar sticker pack.

Solving purchasing is a harder problem to solve, but for the ability to share sticker packs a consistent and predictable format is needed. As part of the user-curated sticker pack work I've been experimenting with I've come up with a very arbitrary JSON format that a given URL could spit out so the sticker picker can self-add the pack. That format looks something like this:
```
{
  "version": 1,
  "type": "m.stickerpack",
  "name": "Huskies",
  "id": "arbitrary_id_for_tracking",
  "author": {
    "type": "mx-user",    // Note: this could also be a company, email, or some other source
    "id": "@travis:t2l.io"
  },
  "license": "CC", // or something
  "stickers": [{
    "id": "arbitrary_id_for_tracking",
    "description": "This is a happy husky",
    "contentUri": "mxc://t2bot.io/b4636e93388542f3cad8fcbb825adf36"
  }]
}
```

As mentioned, there hasn't been much work on my part to further refine this format as it's all experimental. Having sticker packs being compatible between managers (and sticker pickers) is very much needed in my opinion, although solving the paywall problem needs more thought.

File: specification/modules/stickers.rst [hunk 2]
```diff
@@ -36,6 +36,41 @@ Sticker events are received as a single ``m.sticker`` event in the
 
 {{m_sticker_event}}
 
+Integration manager referral URL
+<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+
+In order to aid `widget` or `asset` sharing and discovery, sticker events can
```
Comment (richvdh) [inline specification/modules/stickers.rst]: single backticks don't mean much on their own in RST. not sure what you are going for here?

File: specification/modules/stickers.rst [hunk 3]
```diff
@@ -36,6 +36,41 @@ Sticker events are received as a single ``m.sticker`` event in the
...
+specify an ``integration_manager_url`` property under ``StickerpackInfo``.
+
+This URL is intended to link the user to an appropriate page within the relevant
+Integration Manager to enable them to add / purchase the associated integration
+or integration asset (e.g. a stickerpicker widget, or stickerpack).
+
+The ``integration manager referral URL`` should have the following format::
+
+  https://<hostname>/<integration manager base path>/integration/<integration_type>/<asset_id>?utm_source=<utm_source>&utm_medium=<utm_medium>&utm_campaign=<utm_campaign>&promotional_code=<promotional_code>
```
Comment (richvdh) [inline specification/modules/stickers.rst]: Are the query params and the url path opaque from the client's point of view? If so, why are they specced here? If not, what can it do with them?

File: specification/modules/stickers.rst (hunk 2 above)
Comment (rxl881) [reply specification/modules/stickers.rst]: I think that I was going for emphasis, but it probably just makes sense to remove the formatting altogether.

**Comment by rxl881** [comment]
@turt2live - Thanks for your suggestions on the format for widget "auto-addition". I think that this looks like a really good starting point for self-adding stickerpacks / widgets (and is definitely something that we should build on in due course). However, I think that this is  probably a bit more than what I am trying to achieve here.

The intent of the stickerpack URL as it is at the moment (in this PR) is simply to direct a user clicking on it to a (modal) web-UI for manual addition of the stickerpack (or other asset).

This should work for any Integration Manager (IM) **that is currently configured to work with the user's Matrix client.** E.g. if the sticker pack is from Scalar, the link should open the right page of the scalar web-store, and similarly if it is from Dimension (or other 3rd party IM) the correct page should be opened there. This allows for the correct purchasing & payment flow etc. to be handled by the IM before adding the widget & updating asset records.

Links to integration managers that are not enabled for use by the matrix-client should redirect to an integration manager configuration flow (to add the integration manager in the settings of the client) or a warning that the integration manager is not supported, depending upon the configuration, restrictions or capabilities of the client in use.

It sounds likely that the auto-adding widget/asset link that you propose will be useful as an additional mechanism to this one, probably for cases where no purchasing is required, or once we've established other flows for that, a bit further down the line. In future we might even want to try the auto-adding approach in the first case and then redirect to this endpoint as a fallback, if payment details or manual intervention is required. Does that sound like a plan to you?

**Comment by rxl881** [comment]
@richvdh - I've taken another crack at clarifying the URL format and that the aspects (rendering and tracking query parameters) that matrix clients should be responsible for. Can you please take another look when you have a sec?

**Comment by rxl881** [comment]
@turt2live :
> The behaviour you've described doesn't appeal to me in the sense that it's not inter-operable between different managers.

We need to do more work to support multiple simultaneous integration managers, but I don't see why this isn't inter-operable. Far from it, this allows users to discover stickerpacks and integrations from other integration managers based upon the message content that they see.

**Comment by turt2live** [comment]
I may be misunderstanding something, but I'm not seeing how Scalar can add/edit sticker packs on a Dimension sticker picker (and the other way around)? In your earlier post, it was suggested that the URL open the appropriate integration manager, falling back to a warning if the client can't coordinate the effort. The falling back could easily end up leading to someone purchasing a sticker pack and being unable to use it (or having to make a choice between which picker to use).

**Comment by rxl881** [comment]
We definitely want to get to the point where users can seamlessly have multiple IMs enabled, simultaneously, but there is still some work that needs to be done to get there. This proposed solution allows arbitrary IMs to be used as a stepping-stone in that direction.

> I'm not seeing how Scalar can add/edit sticker packs on a Dimension sticker picker (and the other way around)?

We wouldn't be trying to do this. 

> ... The falling back could easily end up leading to someone purchasing a sticker pack and being unable to use it (or having to make a choice between which picker to use).

This should not be the case. The intention at the moment is for matrix clients to be able to support multiple stickerpicker widgets embedded within the same UI component for the case where users wish to use multiple integration managers simultaneously.

See the comment at - https://github.com/matrix-org/matrix-react-sdk/blob/master/src/components/views/rooms/Stickerpicker.js#L186-L189:

> // Stickers
> // TODO - Add support for Stickerpickers from multiple app stores.
> // Render content from multiple stickerpack sources, each within their
> // own iframe, within the stickerpicker UI element.

From the user's point of view they would still open the sticker selector in the same way, but the AppTile element would load multiple child iframes, each containing a stickerpicker widget (one from scalar and one from Dimension, for example). This should be transparent to the user, other than there being two filter boxes, one at the bottom of each widget, at the moment. We could get round this (to make it more seamless) by having the filter part of the matrix client and extending the postMessage API to send filter update events to the (multiple) sticker pickers, for them to filter and display the correct content.

**Comment by turt2live** [comment]
I still have concerns that this isn't going through the (now formal) proposal process because it's not getting the review it needs. It seems as though client developers, at the least, aren't involved in the discussion here and that can cause problems if this lands in the spec. 

Some clients may not want to implement their sticker picker as multiple iframes. There's already clients out there (Fractal) which are in the early stages of adding stickers, and they don't use iframes at all. Instead, they opt to provide a more native-feeling UI for picking stickers, calling into the API as required (which happens to break when Dimension is used because the API is completely different and not specced for this kind of thing).

tldr: I feel we need to have a larger discussion about how this integration manager stuff is supposed to work and what the use cases are. Adding temporary things to the spec will only make them permanent and hard to remove or replace in the future. Instead, the planning should be put in ahead of time to ensure that use cases aren't missed, and that everyone is on the same page.

(I also have strong feelings, as demonstrated, about how this particular functionality should be addressed)

**Comment by ara4n** [comment]
so the situation here is that we simply haven’t had time to sort the spec here, and rushing it through now outside the proper spec process is well-intended but not going to help. i suggest we shelve it until we have bandwidth again with anyone working on stickers (ie after other priorities are on track on the NV side). in future we’ll try to catch it at the proposal stage (ie before impl).

**Comment by richvdh** [comment]
I actually thought this was an attempt to document the status quo, but it looks as though existing `m.sticker` events lack the `stickerpack` property, so perhaps that's not true?

(the text now looks clear to me, though...)

**Comment by turt2live** [comment]
I'm going to close this while we're waiting for an MSC - the changes now conflict with the current spec and having bitrotted PRs open is not super helpful to us :(

[MSC1951](https://github.com/matrix-org/matrix-doc/pull/1951) lays some foundational work in this area.
